### ISSUE TITLE
Evaluation vs Training (Eval) Accuracy - huge difference.

### ISSUE THREAD

**Issue by tanmayb123** [keras-team/keras#7723]
I've been trying to make a CNN to classify images between 3 classes - `left`, `right`, `forward`. I'm using the following code:

```
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Model
from keras.layers import *
img_width, img_height = 300, 225
train_data_dir = 'data/'
val_data_dir = 'valid/'
train_samples = 7090
val_samples = 600
epoch = 3
# ** Model Begins **
inputs = Input(shape=(img_width, img_height, 3))
seta = Convolution2D(32, 4, 4, activation='relu')(inputs)
seta = MaxPooling2D(3, 3)(seta)
seta = Dropout(0.4)(seta)
seta = Convolution2D(32, 3, 3, activation='relu')(seta)
seta = MaxPooling2D(3, 3)(seta)
seta = Dropout(0.4)(seta)
seta = Convolution2D(32, 2, 2, activation='relu')(seta)
seta = MaxPooling2D(2, 2)(seta)
seta = Dropout(0.4)(seta)
seta = Flatten()(seta)
seta = Dense(70, activation='sigmoid')(seta)
seta = Dense(3, activation='softmax')(seta)
model = Model(input=inputs, output=seta)
# ** Model Ends **
model.summary()
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])
train_datagen = ImageDataGenerator(rescale=1./255)
val_datagen = ImageDataGenerator(rescale=1./255)
train_generator = train_datagen.flow_from_directory(
        train_data_dir,
        target_size=(img_width, img_height),
        batch_size=1,
        class_mode='categorical')
val_generator = val_datagen.flow_from_directory(
        val_data_dir,
        target_size=(img_width, img_height),
        batch_size=1,
        class_mode='categorical')
model.fit_generator(
        train_generator,
        nb_epoch=epoch,
        samples_per_epoch=train_samples,
        class_weight={0: 4, 1: 4, 2: 1.9},
        validation_data=val_generator,
        nb_val_samples=val_samples)
model.save_weights('')
```

The validation accuracy that it reports is `0.9900`. Here's the entire output:

```
7090/7090 [==============================] - 38s - loss: 0.5343 - acc: 0.9426 - val_loss: 0.1117 - val_acc: 0.9900
Epoch 2/3
7090/7090 [==============================] - 38s - loss: 0.2034 - acc: 0.9874 - val_loss: 0.0956 - val_acc: 0.9900
Epoch 3/3
7090/7090 [==============================] - 38s - loss: 0.2003 - acc: 0.9884 - val_loss: 0.0633 - val_acc: 0.9900
```

However, I also have this code, which only checks the validation accuracy on the *exact* same data as the validation data in the training code:
```
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Model
from keras.layers import *
import sys

img_width, img_height = 300, 225

inputs = Input(shape=(img_width, img_height, 3))

seta = Convolution2D(32, 4, 4, activation='relu')(inputs)
seta = MaxPooling2D(3, 3)(seta)
seta = Dropout(0.4)(seta)
seta = Convolution2D(32, 3, 3, activation='relu')(seta)
seta = MaxPooling2D(3, 3)(seta)
seta = Dropout(0.4)(seta)
seta = Convolution2D(32, 2, 2, activation='relu')(seta)
seta = MaxPooling2D(2, 2)(seta)
seta = Dropout(0.4)(seta)
seta = Flatten()(seta)
seta = Dense(70, activation='sigmoid')(seta)
seta = Dense(3, activation='softmax')(seta)

model = Model(input=inputs, output=seta)

model.compile(optimizer='adam', loss='categorical_crossentropy')

val_datagen = ImageDataGenerator(rescale=1./255)

val_generator = val_datagen.flow_from_directory(
        'valid/',
        target_size=(img_width, img_height),
        batch_size=1,
        class_mode='categorical')

val_generator = val_datagen.flow_from_directory(
        'valid/',
        target_size=(img_width, img_height),
        batch_size=1,
        class_mode='categorical')
eval = model.evaluate_generator(val_generator, 600)
print eval
```

It reports the accuracy as:
```
Found 600 images belonging to 3 classes.
0.0595971204911
```

I'm getting 5.9% accuracy with the eval script, and 99% accuracy with the eval-in-training script! Am I doing something wrong?

**Comment by kalluwa** [comment]
I met the same problem but on training set, :( 
btw, i didn't see any weight loading function in your code

**Comment by tanmayb123** [comment]
Oh sorry, my code **does** do that. I removed one line once I pasted it in.

**Comment by juggernaut5k** [comment]
I believe you want to use categorical_accuracy for your metrics instead of accuracy. I've experienced that one in the past. Accuracy is for non one-hot encoded data.
